File: src/tools/finance/beta-hmm.ts
/**
 * Beta-HMM — Hidden Markov Model with Beta emissions for bounded [0,1] data.
 *
 * Designed for Polymarket prices (probability series) which Gaussian / Student-t
 * emissions cannot represent natively (unbounded support, no U/J shapes).
 *
 * Reference: Voigt (2025), *Predicting Prediction Markets: A Beta-Hidden Markov
 * Modeling Approach* — see references/prediction-markets/BetaHMMpolymarket-18-1.pdf.
 *
 * Algorithm:
 *   - Emission family: Beta(αᵢ, βᵢ) per state i ∈ {0..K-1}.
 *   - Forward / Backward / Baum-Welch identical in shape to Gaussian HMM,
 *     replacing gaussianPdf with betaPdf.
 *   - M-step refits {αᵢ, βᵢ} via *weighted* method-of-moments (closed form)
 *     using the posterior γ[t][i] as weights. MoM is the standard substitute
 *     for the analytically-intractable Beta MLE in HMM EM updates.
 *
 * This module mirrors the Python implementation in research/models/beta_hmm.py.
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface BetaEmission {
  alpha: number;
  beta: number;
}

export interface BetaHMMParams {
  nStates: number;
  pi: number[];
  A: number[][];
  emissions: BetaEmission[];
}

export interface BetaHMMFitResult {
  params: BetaHMMParams;
  logLikelihood: number;
  iterations: number;
  converged: boolean;
}

// ---------------------------------------------------------------------------
// log-Gamma (Lanczos) — duplicated from markov-distribution.ts to keep this
// module self-contained.
// ---------------------------------------------------------------------------

function lgamma(x: number): number {
  if (x < 0.5) return Math.log(Math.PI / Math.sin(Math.PI * x)) - lgamma(1 - x);
  const g = 7;
  const c = [
    0.99999999999980993, 676.5203681218851, -1259.1392167224028,
    771.32342877765313, -176.61502916214059, 12.507343278686905,
    -0.13857109526572012, 9.9843695780195716e-6, 1.5056327351493116e-7,
  ];
  x -= 1;
  let a = c[0];
  for (let i = 1; i < g + 2; i++) a += c[i] / (x + i);
  const t = x + g + 0.5;
  return 0.5 * Math.log(2 * Math.PI) + (x + 0.5) * Math.log(t) - t + Math.log(a);
}

// ---------------------------------------------------------------------------
// Beta PDF
// ---------------------------------------------------------------------------

/** Beta probability density function. Returns 0 outside (0,1) for safety. */
export function betaPdf(x: number, e: BetaEmission): number {
  if (x <= 0 || x >= 1) return 0;
  if (e.alpha <= 0 || e.beta <= 0) return 0;
  const lnB = lgamma(e.alpha) + lgamma(e.beta) - lgamma(e.alpha + e.beta);
  const logPdf = (e.alpha - 1) * Math.log(x) + (e.beta - 1) * Math.log(1 - x) - lnB;
  // Clamp ridiculously large values for numerical safety.
  return Math.exp(Math.max(-700, Math.min(700, logPdf)));
}

// ---------------------------------------------------------------------------
// Method-of-moments fit (weighted)
// ---------------------------------------------------------------------------

/**
 * Weighted method-of-moments fit for Beta(α,β) given samples and per-sample
 * weights (typically the HMM posterior γ[t][i] for state i).
 *
 * MoM equations:
 *   m = Σwx / Σw                     (weighted mean)
 *   v = Σw(x - m)² / Σw              (weighted variance, MLE form)
 *   ν = m(1 - m)/v − 1               (effective sample size)
 *   α = ν · m,   β = ν · (1 - m)
 *
 * Falls back to Beta(1,1) (uniform) when variance is non-positive or weights
 * vanish. Samples are clipped to [ε, 1−ε] so degenerate boundary observations
 * cannot collapse the variance.
 */
export function fitBetaMoM(weights: number[], samples: number[]): BetaEmission {
  if (weights.length !== samples.length) {
    throw new Error('fitBetaMoM: weights and samples must have equal length');
  }
  const eps = 1e-6;
  let wsum = 0;
  let wxsum = 0;
  for (let i = 0; i < samples.length; i++) {
    const w = weights[i];
    if (!Number.isFinite(w) || w <= 0) continue;
    const x = Math.min(1 - eps, Math.max(eps, samples[i]));
    wsum += w;
    wxsum += w * x;
  }
  if (wsum <= 0) return { alpha: 1, beta: 1 };
  const m = wxsum / wsum;

  let wvsum = 0;
  for (let i = 0; i < samples.length; i++) {
    const w = weights[i];
    if (!Number.isFinite(w) || w <= 0) continue;
    const x = Math.min(1 - eps, Math.max(eps, samples[i]));
    wvsum += w * (x - m) * (x - m);
  }
  const v = wvsum / wsum;
  if (!(v > 1e-10) || m <= 0 || m >= 1) return { alpha: 1, beta: 1 };

  const nu = (m * (1 - m)) / v - 1;
  if (!(nu > 0)) return { alpha: 1, beta: 1 };

  const alpha = nu * m;
  const beta = nu * (1 - m);
  if (!Number.isFinite(alpha) || !Number.isFinite(beta) || alpha <= 0 || beta <= 0) {
    return { alpha: 1, beta: 1 };
  }
  return { alpha, beta };
}

// ---------------------------------------------------------------------------
// Initialisation
// ---------------------------------------------------------------------------

/**
 * Initialise a Beta-HMM by quantile-splitting the observations and fitting
 * Beta(α,β) to each bucket via method-of-moments. Transition matrix starts
 * mildly sticky (0.85 self-loop). Initial state π is uniform.
 */
export function initializeBetaHMM(observations: number[], nStates: number): BetaHMMParams {
  if (nStates < 1) throw new Error('initializeBetaHMM: nStates must be >= 1');
  const sorted = [...observations].sort((a, b) => a - b);
  const n = sorted.length;
  const emissions: BetaEmission[] = [];
  for (let i = 0; i < nStates; i++) {
    const lo = Math.floor((i * n) / nStates);
    const hi = Math.floor(((i + 1) * n) / nStates);
    const slice = sorted.slice(lo, Math.max(hi, lo + 1));
    const w = slice.map(() => 1);
    emissions.push(fitBetaMoM(w, slice));
  }
  const stay = 0.85;
  const off = (1 - stay) / Math.max(nStates - 1, 1);
  const A: number[][] = [];
  for (let i = 0; i < nStates; i++) {
    A.push(Array.from({ length: nStates }, (_, j) => (i === j ? stay : off)));
  }
  const pi = Array.from({ length: nStates }, () => 1 / nStates);
  return { nStates, pi, A, emissions };
}

// ---------------------------------------------------------------------------
// Forward / Backward
// ---------------------------------------------------------------------------

export function forwardBeta(obs: number[], p: BetaHMMParams): { alpha: number[][]; scales: number[]; logLik: number } {
  const T = obs.length;
  const K = p.nStates;
  const alpha: number[][] = Array.from({ length: T }, () => Array(K).fill(0));
  const scales: number[] = Array(T).fill(0);

  // t = 0
  let s0 = 0;
  for (let i = 0; i < K; i++) {
    alpha[0][i] = p.pi[i] * Math.max(betaPdf(obs[0], p.emissions[i]), 1e-300);
    s0 += alpha[0][i];
  }
  scales[0] = s0 || 1;
  for (let i = 0; i < K; i++) alpha[0][i] /= scales[0];

  for (let t = 1; t < T; t++) {
    let st = 0;
    for (let j = 0; j < K; j++) {
      let acc = 0;
      for (let i = 0; i < K; i++) acc += alpha[t - 1][i] * p.A[i][j];
      alpha[t][j] = acc * Math.max(betaPdf(obs[t], p.emissions[j]), 1e-300);
      st += alpha[t][j];
    }
    scales[t] = st || 1;
    for (let j = 0; j < K; j++) alpha[t][j] /= scales[t];
  }

  let logLik = 0;
  for (let t = 0; t < T; t++) logLik += Math.log(scales[t]);
  return { alpha, scales, logLik };
}

export function backwardBeta(obs: number[], p: BetaHMMParams, scales: number[]): number[][] {
  const T = obs.length;
  const K = p.nStates;
  const beta: number[][] = Array.from({ length: T }, () => Array(K).fill(0));
  for (let i = 0; i < K; i++) beta[T - 1][i] = 1 / scales[T - 1];
  for (let t = T - 2; t >= 0; t--) {
    for (let i = 0; i < K; i++) {
      let acc = 0;
      for (let j = 0; j < K; j++) {
        acc += p.A[i][j] * Math.max(betaPdf(obs[t + 1], p.emissions[j]), 1e-300) * beta[t + 1][j];
      }
      beta[t][i] = acc / scales[t];
    }
  }
  return beta;
}

// ---------------------------------------------------------------------------
// Baum-Welch
// ---------------------------------------------------------------------------

export function baumWelchBeta(
  obs: number[],
  init: BetaHMMParams,
  maxIter = 50,
  tol = 1e-4,
): BetaHMMFitResult {
  let p = clone(init);
  const T = obs.length;
  const K = p.nStates;
  let prevLL = -Infinity;
  let converged = false;
  let iter = 0;
  let lastLL = -Infinity;

  for (iter = 0; iter < maxIter; iter++) {
    const { alpha, scales, logLik } = forwardBeta(obs, p);
    const beta = backwardBeta(obs, p, scales);
    lastLL = logLik;

    // γ[t][i] = α[t][i] β[t][i] · scales[t]   (because of scaling normalisation)
    const gamma: number[][] = Array.from({ length: T }, () => Array(K).fill(0));
    for (let t = 0; t < T; t++) {
      let s = 0;
      for (let i = 0; i < K; i++) {
        gamma[t][i] = alpha[t][i] * beta[t][i] * scales[t];
        s += gamma[t][i];
      }
      if (s > 0) for (let i = 0; i < K; i++) gamma[t][i] /= s;
    }

    // ξ[t][i][j] expected joint state pairs
    const xiSum: number[][] = Array.from({ length: K }, () => Array(K).fill(0));
    for (let t = 0; t < T - 1; t++) {
      let denom = 0;
      const xi_t: number[][] = Array.from({ length: K }, () => Array(K).fill(0));
      for (let i = 0; i < K; i++) {
        for (let j = 0; j < K; j++) {
          xi_t[i][j] = alpha[t][i] * p.A[i][j]
            * Math.max(betaPdf(obs[t + 1], p.emissions[j]), 1e-300)
            * beta[t + 1][j];
          denom += xi_t[i][j];
        }
      }
      if (denom > 0) {
        for (let i = 0; i < K; i++) for (let j = 0; j < K; j++) xiSum[i][j] += xi_t[i][j] / denom;
      }
    }

    // M-step
    const piNew = gamma[0].slice();
    const piSum = piNew.reduce((s, v) => s + v, 0) || 1;
    for (let i = 0; i < K; i++) piNew[i] /= piSum;

    const Anew: number[][] = Array.from({ length: K }, () => Array(K).fill(0));
    for (let i = 0; i < K; i++) {
      const denom = xiSum[i].reduce((s, v) => s + v, 0);
      if (denom > 0) {
        for (let j = 0; j < K; j++) Anew[i][j] = xiSum[i][j] / denom;
      } else {
        // Stay-state fallback
        Anew[i] = Array.from({ length: K }, (_, j) => (i === j ? 1 : 0));
      }
    }

    const emissionsNew: BetaEmission[] = [];
    for (let i = 0; i < K; i++) {
      const w = gamma.map(row => row[i]);
      emissionsNew.push(fitBetaMoM(w, obs));
    }

    p = { nStates: K, pi: piNew, A: Anew, emissions: emissionsNew };

    if (Math.abs(logLik - prevLL) < tol) {
      converged = true;
      break;
    }
    prevLL = logLik;
  }

  return { params: p, logLikelihood: lastLL, iterations: iter + 1, converged };
}

// ---------------------------------------------------------------------------
// Viterbi (most-likely state path)
// ---------------------------------------------------------------------------

export function viterbiBeta(obs: number[], p: BetaHMMParams): number[] {
  const T = obs.length;
  const K = p.nStates;
  const delta: number[][] = Array.from({ length: T }, () => Array(K).fill(-Infinity));
  const psi: number[][] = Array.from({ length: T }, () => Array(K).fill(0));

  for (let i = 0; i < K; i++) {
    delta[0][i] = Math.log(Math.max(p.pi[i], 1e-300))
      + Math.log(Math.max(betaPdf(obs[0], p.emissions[i]), 1e-300));
  }

  for (let t = 1; t < T; t++) {
    for (let j = 0; j < K; j++) {
      let bestVal = -Infinity;
      let bestI = 0;
      for (let i = 0; i < K; i++) {
        const val = delta[t - 1][i] + Math.log(Math.max(p.A[i][j], 1e-300));
        if (val > bestVal) {
          bestVal = val;
          bestI = i;
        }
      }
      delta[t][j] = bestVal + Math.log(Math.max(betaPdf(obs[t], p.emissions[j]), 1e-300));
      psi[t][j] = bestI;
    }
  }

  const path: number[] = Array(T).fill(0);
  let bestFinal = 0;
  let bestVal = -Infinity;
  for (let i = 0; i < K; i++) {
    if (delta[T - 1][i] > bestVal) {
      bestVal = delta[T - 1][i];
      bestFinal = i;
    }
  }
  path[T - 1] = bestFinal;
  for (let t = T - 2; t >= 0; t--) path[t] = psi[t + 1][path[t + 1]];
  return path;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function clone(p: BetaHMMParams): BetaHMMParams {
  return {
    nStates: p.nStates,
    pi: p.pi.slice(),
    A: p.A.map(row => row.slice()),
    emissions: p.emissions.map(e => ({ ...e })),
  };
}

File: research/models/beta_hmm.py
"""Beta-HMM — Hidden Markov Model with Beta emissions for bounded [0,1] data.

Mirrors src/tools/finance/beta-hmm.ts. Designed for Polymarket prices.

Reference: Voigt (2025), *Predicting Prediction Markets: A Beta-Hidden Markov
Modeling Approach* — see references/prediction-markets/BetaHMMpolymarket-18-1.pdf.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Sequence

import math

import numpy as np
from scipy.special import gammaln

# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------


@dataclass
class BetaEmission:
    alpha: float
    beta: float


@dataclass
class BetaHMMParams:
    n_states: int
    pi: np.ndarray  # shape (K,)
    A: np.ndarray   # shape (K, K)
    emissions: List[BetaEmission] = field(default_factory=list)


@dataclass
class BetaHMMFitResult:
    params: BetaHMMParams
    log_likelihood: float
    iterations: int
    converged: bool


# ---------------------------------------------------------------------------
# Beta PDF
# ---------------------------------------------------------------------------


def beta_pdf(x: float, e: BetaEmission) -> float:
    """Beta probability density. Returns 0 outside (0,1)."""
    if x <= 0 or x >= 1:
        return 0.0
    if e.alpha <= 0 or e.beta <= 0:
        return 0.0
    ln_b = gammaln(e.alpha) + gammaln(e.beta) - gammaln(e.alpha + e.beta)
    log_pdf = (e.alpha - 1) * math.log(x) + (e.beta - 1) * math.log(1 - x) - ln_b
    return math.exp(max(-700.0, min(700.0, log_pdf)))


# ---------------------------------------------------------------------------
# Method-of-moments fit (weighted)
# ---------------------------------------------------------------------------


def fit_beta_mom(weights: Sequence[float], samples: Sequence[float]) -> BetaEmission:
    """Weighted method-of-moments Beta(α,β) fit.

    See TS docstring for equations. Falls back to Beta(1,1) if variance ≤ 0.
    """
    if len(weights) != len(samples):
        raise ValueError("fit_beta_mom: weights and samples must have equal length")
    eps = 1e-6
    w = np.asarray(weights, dtype=float)
    x = np.clip(np.asarray(samples, dtype=float), eps, 1 - eps)
    valid = np.isfinite(w) & (w > 0)
    if not valid.any():
        return BetaEmission(1.0, 1.0)
    w = w[valid]
    x = x[valid]
    wsum = float(w.sum())
    if wsum <= 0:
        return BetaEmission(1.0, 1.0)
    m = float((w * x).sum() / wsum)
    v = float((w * (x - m) ** 2).sum() / wsum)
    if v <= 1e-10 or m <= 0 or m >= 1:
        return BetaEmission(1.0, 1.0)
    nu = m * (1 - m) / v - 1
    if nu <= 0:
        return BetaEmission(1.0, 1.0)
    alpha = nu * m
    beta = nu * (1 - m)
    if not (math.isfinite(alpha) and math.isfinite(beta)) or alpha <= 0 or beta <= 0:
        return BetaEmission(1.0, 1.0)
    return BetaEmission(float(alpha), float(beta))


# ---------------------------------------------------------------------------
# Initialisation
# ---------------------------------------------------------------------------


def initialize_beta_hmm(observations: Sequence[float], n_states: int) -> BetaHMMParams:
    if n_states < 1:
        raise ValueError("initialize_beta_hmm: n_states must be >= 1")
    sorted_obs = np.sort(np.asarray(observations, dtype=float))
    n = len(sorted_obs)
    emissions: List[BetaEmission] = []
    for i in range(n_states):
        lo = (i * n) // n_states
        hi = max(((i + 1) * n) // n_states, lo + 1)
        sl = sorted_obs[lo:hi]
        emissions.append(fit_beta_mom([1.0] * len(sl), sl.tolist()))
    stay = 0.85
    off = (1 - stay) / max(n_states - 1, 1)
    A = np.full((n_states, n_states), off)
    np.fill_diagonal(A, stay)
    pi = np.full(n_states, 1.0 / n_states)
    return BetaHMMParams(n_states=n_states, pi=pi, A=A, emissions=emissions)


# ---------------------------------------------------------------------------
# Forward / Backward
# ---------------------------------------------------------------------------


def _emit_vec(obs_t: float, p: BetaHMMParams) -> np.ndarray:
    return np.array([max(beta_pdf(obs_t, e), 1e-300) for e in p.emissions])


def forward_beta(obs: Sequence[float], p: BetaHMMParams):
    T = len(obs)
    K = p.n_states
    alpha = np.zeros((T, K))
    scales = np.zeros(T)
    e0 = _emit_vec(obs[0], p)
    alpha[0] = p.pi * e0
    scales[0] = alpha[0].sum() or 1.0
    alpha[0] /= scales[0]
    for t in range(1, T):
        e_t = _emit_vec(obs[t], p)
        alpha[t] = (alpha[t - 1] @ p.A) * e_t
        scales[t] = alpha[t].sum() or 1.0
        alpha[t] /= scales[t]
    log_lik = float(np.log(scales).sum())
    return alpha, scales, log_lik


def backward_beta(obs: Sequence[float], p: BetaHMMParams, scales: np.ndarray) -> np.ndarray:
    T = len(obs)
    K = p.n_states
    beta = np.zeros((T, K))
    beta[-1] = 1.0 / scales[-1]
    for t in range(T - 2, -1, -1):
        e_next = _emit_vec(obs[t + 1], p)
        beta[t] = (p.A @ (e_next * beta[t + 1])) / scales[t]
    return beta


# ---------------------------------------------------------------------------
# Baum-Welch
# ---------------------------------------------------------------------------


def baum_welch_beta(
    obs: Sequence[float],
    init: BetaHMMParams,
    max_iter: int = 50,
    tol: float = 1e-4,
) -> BetaHMMFitResult:
    p = BetaHMMParams(
        n_states=init.n_states,
        pi=init.pi.copy(),
        A=init.A.copy(),
        emissions=[BetaEmission(e.alpha, e.beta) for e in init.emissions],
    )
    obs_arr = np.asarray(obs, dtype=float)
    T = len(obs_arr)
    K = p.n_states
    prev_ll = -math.inf
    converged = False
    last_ll = -math.inf
    iter_count = 0
    for iter_count in range(max_iter):
        alpha, scales, log_lik = forward_beta(obs_arr, p)
        beta = backward_beta(obs_arr, p, scales)
        last_ll = log_lik

        gamma = alpha * beta * scales[:, None]
        row_sums = gamma.sum(axis=1, keepdims=True)
        row_sums[row_sums == 0] = 1.0
        gamma = gamma / row_sums

        xi_sum = np.zeros((K, K))
        for t in range(T - 1):
            e_next = _emit_vec(obs_arr[t + 1], p)
            xi_t = (alpha[t][:, None] * p.A) * (e_next * beta[t + 1])[None, :]
            denom = xi_t.sum()
            if denom > 0:
                xi_sum += xi_t / denom

        pi_new = gamma[0].copy()
        s = pi_new.sum() or 1.0
        pi_new /= s

        A_new = np.zeros((K, K))
        for i in range(K):
            denom = xi_sum[i].sum()
            if denom > 0:
                A_new[i] = xi_sum[i] / denom
            else:
                A_new[i, i] = 1.0

        emissions_new: List[BetaEmission] = []
        for i in range(K):
            emissions_new.append(fit_beta_mom(gamma[:, i].tolist(), obs_arr.tolist()))

        p = BetaHMMParams(n_states=K, pi=pi_new, A=A_new, emissions=emissions_new)

        if abs(log_lik - prev_ll) < tol:
            converged = True
            break
        prev_ll = log_lik

    return BetaHMMFitResult(params=p, log_likelihood=last_ll, iterations=iter_count + 1, converged=converged)


# ---------------------------------------------------------------------------
# Viterbi
# ---------------------------------------------------------------------------


def viterbi_beta(obs: Sequence[float], p: BetaHMMParams) -> List[int]:
    T = len(obs)
    K = p.n_states
    delta = np.full((T, K), -math.inf)
    psi = np.zeros((T, K), dtype=int)
    e0 = _emit_vec(obs[0], p)
    delta[0] = np.log(np.maximum(p.pi, 1e-300)) + np.log(np.maximum(e0, 1e-300))
    log_A = np.log(np.maximum(p.A, 1e-300))
    for t in range(1, T):
        e_t = _emit_vec(obs[t], p)
        for j in range(K):
            vals = delta[t - 1] + log_A[:, j]
            best_i = int(np.argmax(vals))
            delta[t, j] = vals[best_i] + math.log(max(e_t[j], 1e-300))
            psi[t, j] = best_i
    path = [0] * T
    path[-1] = int(np.argmax(delta[-1]))
    for t in range(T - 2, -1, -1):
        path[t] = int(psi[t + 1, path[t + 1]])
    return path
